### ISSUE TITLE
ansible inline vault

### ISSUE THREAD

**Issue by h0nIg** [ansible/ansible#13287]
Sometimes you want to define a complex variable dictionary and do not want to create a separate vault file for passwords. It would be cool for history purpose to use a plain yml file for variable definition and inline vault variable content like the following. This feature is described by @jpmens on http://jpmens.net/2014/02/22/my-thoughts-on-ansible-s-vault/

```
nameserver: 10.0.0.1
admin: Jane Jolie
dbpassword: @vault@"AHchx0a+G8mejs84tGxCNKxMFP7tM7Y7kl"
webservertype: nginx
```

**Comment by bcoca** [comment]
closing as dupe of #12087

**Comment by h0nIg** [comment]
this is not a duplicate, since the other guy wants to get this implemented by filters and my intention is that this is handled by the ansible yaml parser itself, please reopen!

**Comment by bcoca** [comment]
i doubt that will happen as the performance penalty of examining every possible value for vault content will slow down Ansible considerably.

**Comment by h0nIg** [comment]
therefore the initial idea was to scan only the first letters and check for a special content like @vault@

**Comment by bcoca** [comment]
yes, that is what will be prohibitive for most users, only those with tiny inventories and few variables would not notice a huge delay.

We have been doing a lot of performance optimization prior to the 2.0 release, most speedups came from avoiding copying and examining the variables as much as we did.

**Comment by h0nIg** [comment]
hmm ok.

but to check this statement: do you really think another loop on VariableManager.get_vars for replacing variable content is expensive too? (in comparision to copying on combine_vars). just scrolled the source code and it seems that it would work

**Comment by bcoca** [comment]
try it out, but look at the commit history and you'll see we removed lots of code that scanned and/or copied the variables. What you see now should be the minimal needed for ansible's functionality and it is still slower than what we would want when large inventories are invovled

**Comment by bcoca** [comment]
@h0nIg i do think your idea makes for a better user interface, though no need to add @vault@ as vault already puts in a specific recognizable header.

If you can implement w/o it impacting performance too much, go ahead, i suspect it will be hard to do but will be very happy if you prove me wrong.

**Comment by bcoca** [comment]
i've been thinking about this more, a much more efficient way to do this would be to use a YAML type handler instead of parsing the values.

```
key: !vault vaultedvalue
```

would just require that we create a function that handles vaulted strings and register it with the module parser as the one to handle the `!vault` type.

**Comment by h0nIg** [comment]
it seems this is pretty easy, first implementation (not tested)

```
diff --git a/lib/ansible/parsing/dataloader.py b/lib/ansible/parsing/dataloader.py
index c54ba78..6f2c2fc 100644
--- a/lib/ansible/parsing/dataloader.py
+++ b/lib/ansible/parsing/dataloader.py
@@ -143,7 +143,7 @@ class DataLoader():
     def _safe_load(self, stream, file_name=None):
         ''' Implements yaml.safe_load(), except using our custom loader class. '''

-        loader = AnsibleLoader(stream, file_name)
+        loader = AnsibleLoader(self, stream, file_name)
         try:
             return loader.get_single_data()
         finally:
diff --git a/lib/ansible/parsing/yaml/constructor.py b/lib/ansible/parsing/yaml/constructor.py
index bac0d3f..8174529 100644
--- a/lib/ansible/parsing/yaml/constructor.py
+++ b/lib/ansible/parsing/yaml/constructor.py
@@ -31,7 +31,8 @@ except ImportError:


 class AnsibleConstructor(Constructor):
-    def __init__(self, file_name=None):
+    def __init__(self, loader, file_name=None):
+        self._loader = loader
         self._ansible_file_name = file_name
         super(AnsibleConstructor, self).__init__()

@@ -102,6 +103,14 @@ class AnsibleConstructor(Constructor):

         return (datasource, line, column)

+    def construct_vault(self, node):
+        value = loader.construct_scalar(node)
+        try:
+            return self._loader.load(value)
+        except AnsibleParserError as exc:
+            raise ConstructorError(None, None,
+                    "found invalid vault string", node.start_mark)
+
 AnsibleConstructor.add_constructor(
     u'tag:yaml.org,2002:map',
     AnsibleConstructor.construct_yaml_map)
@@ -121,3 +130,7 @@ AnsibleConstructor.add_constructor(
 AnsibleConstructor.add_constructor(
     u'tag:yaml.org,2002:seq',
     AnsibleConstructor.construct_yaml_seq)
+
+AnsibleConstructor.add_constructor(
+    u'!vault',
+    AnsibleConstructor.construct_vault)
\ No newline at end of file
diff --git a/lib/ansible/parsing/yaml/loader.py b/lib/ansible/parsing/yaml/loader.py
index e8547ff..5eb804d 100644
--- a/lib/ansible/parsing/yaml/loader.py
+++ b/lib/ansible/parsing/yaml/loader.py
@@ -31,9 +31,9 @@ from ansible.parsing.yaml.constructor import AnsibleConstructor

 if HAVE_PYYAML_C:
     class AnsibleLoader(CParser, AnsibleConstructor, Resolver):
-        def __init__(self, stream, file_name=None):
+        def __init__(self, loader, stream, file_name=None):
             CParser.__init__(self, stream)
-            AnsibleConstructor.__init__(self, file_name=file_name)
+            AnsibleConstructor.__init__(self, loader, file_name=file_name)
             Resolver.__init__(self)
 else:
     from yaml.composer import Composer
@@ -42,10 +42,10 @@ else:
     from yaml.parser import Parser

     class AnsibleLoader(Reader, Scanner, Parser, Composer, AnsibleConstructor, Resolver):
-        def __init__(self, stream, file_name=None):
+        def __init__(self, loader, stream, file_name=None):
             Reader.__init__(self, stream)
             Scanner.__init__(self)
             Parser.__init__(self)
             Composer.__init__(self)
-            AnsibleConstructor.__init__(self, file_name=file_name)
+            AnsibleConstructor.__init__(self, loader, file_name=file_name)
             Resolver.__init__(self)
```

**Comment by h0nIg** [comment]
this seems to be pretty easy, wrote example implementation (not tested):

diff --git a/lib/ansible/parsing/dataloader.py b/lib/ansible/parsing/dataloader.py
index c54ba78..6d70a0d 100644
--- a/lib/ansible/parsing/dataloader.py
+++ b/lib/ansible/parsing/dataloader.py
@@ -143,7 +143,7 @@ class DataLoader():
     def _safe_load(self, stream, file_name=None):
         ''' Implements yaml.safe_load(), except using our custom loader class. '''

-        loader = AnsibleLoader(stream, file_name)
+        loader = AnsibleLoader(stream, file_name, self._vault_password)
         try:
             return loader.get_single_data()
         finally:
diff --git a/lib/ansible/parsing/yaml/constructor.py b/lib/ansible/parsing/yaml/constructor.py
index bac0d3f..2de59e5 100644
--- a/lib/ansible/parsing/yaml/constructor.py
+++ b/lib/ansible/parsing/yaml/constructor.py
@@ -22,6 +22,7 @@ __metaclass__ = type
 from yaml.constructor import Constructor, ConstructorError
 from yaml.nodes import MappingNode
 from ansible.parsing.yaml.objects import AnsibleMapping, AnsibleSequence, AnsibleUnicode
+from ansible.parsing.vault import VaultLib

 try:
     from __main__ import display
@@ -31,7 +32,8 @@ except ImportError:


 class AnsibleConstructor(Constructor):
-    def __init__(self, file_name=None):
+    def __init__(self, vault_password, file_name=None):
+        self._vault_password = vault_password
         self._ansible_file_name = file_name
         super(AnsibleConstructor, self).__init__()

@@ -102,6 +104,23 @@ class AnsibleConstructor(Constructor):

         return (datasource, line, column)

+    def construct_vault(self, node):
+        data = loader.construct_scalar(node)
+
+        vault = VaultLib(password=self._vault_password)
+        if not vault.is_encrypted(data):
+            raise ConstructorError(None, None,
+                    "found vault but argument is not encrypted", node.start_mark)
+
+        data = vault.decrypt(data)
+        loader = AnsibleLoader(data)
+
+        try:
+            return loader.get_single_data()
+        except YAMLError as exc:
+            raise ConstructorError(None, None,
+                    "found valid vault string but content is invalid", node.start_mark)
+
 AnsibleConstructor.add_constructor(
     u'tag:yaml.org,2002:map',
     AnsibleConstructor.construct_yaml_map)
@@ -121,3 +140,7 @@ AnsibleConstructor.add_constructor(
 AnsibleConstructor.add_constructor(
     u'tag:yaml.org,2002:seq',
     AnsibleConstructor.construct_yaml_seq)
+
+AnsibleConstructor.add_constructor(
+    u'!vault',
+    AnsibleConstructor.construct_vault)
\ No newline at end of file
diff --git a/lib/ansible/parsing/yaml/loader.py b/lib/ansible/parsing/yaml/loader.py
index e8547ff..fe48e9f 100644
--- a/lib/ansible/parsing/yaml/loader.py
+++ b/lib/ansible/parsing/yaml/loader.py
@@ -31,9 +31,9 @@ from ansible.parsing.yaml.constructor import AnsibleConstructor

 if HAVE_PYYAML_C:
     class AnsibleLoader(CParser, AnsibleConstructor, Resolver):
-        def __init__(self, stream, file_name=None):
+        def __init__(self, stream, file_name=None, vault_password=None):
             CParser.__init__(self, stream)
-            AnsibleConstructor.__init__(self, file_name=file_name)
+            AnsibleConstructor.__init__(self, file_name=file_name, vault_password=vault_password)
             Resolver.__init__(self)
 else:
     from yaml.composer import Composer
@@ -42,10 +42,10 @@ else:
     from yaml.parser import Parser

     class AnsibleLoader(Reader, Scanner, Parser, Composer, AnsibleConstructor, Resolver):
-        def __init__(self, stream, file_name=None):
+        def __init__(self, stream, file_name=None, vault_password=None):
             Reader.__init__(self, stream)
             Scanner.__init__(self)
             Parser.__init__(self)
             Composer.__init__(self)
-            AnsibleConstructor.__init__(self, file_name=file_name)
+            AnsibleConstructor.__init__(self, file_name=file_name, vault_password=vault_password)
             Resolver.__init__(self)

**Comment by h0nIg** [comment]
second example: delegate implementation (supports json too, but it might be not a good idea to forward the dataloader object)

```
diff --git a/lib/ansible/parsing/dataloader.py b/lib/ansible/parsing/dataloader.py
index c54ba78..6f2c2fc 100644
--- a/lib/ansible/parsing/dataloader.py
+++ b/lib/ansible/parsing/dataloader.py
@@ -143,7 +143,7 @@ class DataLoader():
     def _safe_load(self, stream, file_name=None):
         ''' Implements yaml.safe_load(), except using our custom loader class. '''

-        loader = AnsibleLoader(stream, file_name)
+        loader = AnsibleLoader(self, stream, file_name)
         try:
             return loader.get_single_data()
         finally:
diff --git a/lib/ansible/parsing/yaml/constructor.py b/lib/ansible/parsing/yaml/constructor.py
index bac0d3f..8174529 100644
--- a/lib/ansible/parsing/yaml/constructor.py
+++ b/lib/ansible/parsing/yaml/constructor.py
@@ -31,7 +31,8 @@ except ImportError:


 class AnsibleConstructor(Constructor):
-    def __init__(self, file_name=None):
+    def __init__(self, loader, file_name=None):
+        self._loader = loader
         self._ansible_file_name = file_name
         super(AnsibleConstructor, self).__init__()

@@ -102,6 +103,14 @@ class AnsibleConstructor(Constructor):

         return (datasource, line, column)

+    def construct_vault(self, node):
+        data = loader.construct_scalar(node)
+        try:
+            return self._loader.load(data)
+        except AnsibleParserError as exc:
+            raise ConstructorError(None, None,
+                    "found invalid vault string", node.start_mark)
+
 AnsibleConstructor.add_constructor(
     u'tag:yaml.org,2002:map',
     AnsibleConstructor.construct_yaml_map)
@@ -121,3 +130,7 @@ AnsibleConstructor.add_constructor(
 AnsibleConstructor.add_constructor(
     u'tag:yaml.org,2002:seq',
     AnsibleConstructor.construct_yaml_seq)
+
+AnsibleConstructor.add_constructor(
+    u'!vault',
+    AnsibleConstructor.construct_vault)
\ No newline at end of file
diff --git a/lib/ansible/parsing/yaml/loader.py b/lib/ansible/parsing/yaml/loader.py
index e8547ff..5eb804d 100644
--- a/lib/ansible/parsing/yaml/loader.py
+++ b/lib/ansible/parsing/yaml/loader.py
@@ -31,9 +31,9 @@ from ansible.parsing.yaml.constructor import AnsibleConstructor

 if HAVE_PYYAML_C:
     class AnsibleLoader(CParser, AnsibleConstructor, Resolver):
-        def __init__(self, stream, file_name=None):
+        def __init__(self, loader, stream, file_name=None):
             CParser.__init__(self, stream)
-            AnsibleConstructor.__init__(self, file_name=file_name)
+            AnsibleConstructor.__init__(self, loader, file_name=file_name)
             Resolver.__init__(self)
 else:
     from yaml.composer import Composer
@@ -42,10 +42,10 @@ else:
     from yaml.parser import Parser

     class AnsibleLoader(Reader, Scanner, Parser, Composer, AnsibleConstructor, Resolver):
-        def __init__(self, stream, file_name=None):
+        def __init__(self, loader, stream, file_name=None):
             Reader.__init__(self, stream)
             Scanner.__init__(self)
             Parser.__init__(self)
             Composer.__init__(self)
-            AnsibleConstructor.__init__(self, file_name=file_name)
+            AnsibleConstructor.__init__(self, loader, file_name=file_name)
             Resolver.__init__(self)

```

**Comment by h0nIg** [comment]
this seems to be pretty easy, wrote example implementation (not tested):

https://github.com/h0nIg/ansible/commit/e680f504aa0f6826ff47263d79dfb19d8a74b0e7

**Comment by marcinhlybin** [comment]
I wrote ansible filter that does exactly what you want:
https://github.com/ahes/ansible-filter-plugins/blob/master/vault.py

Have fun :)

**Comment by h0nIg** [comment]
@ahes i do not want to include a separate encryption library, i want to use it with ansible capabilities

**Comment by alikins** [comment]
http://jpmens.net/2014/02/22/my-thoughts-on-ansible-s-vault/   describes something that decrypts to a string. 

Would an inline vault need to support decrypting to a more complex object (map, list, etc)?

If a inline vault needed additional info (like an identifier for which key to use), would it be acceptable if that was in a wrapper mapping?  Or 'args' to a !vault object?

**Comment by h0nIg** [comment]
@alikins 

the first one is the intended one.

inline vault should not need to decrypt to a complex object, but it should support to decrypt to a yaml multiline string. the intention is to hide the parts that should be secure **only**

what do you mean by "wrapper mapping"? i do not have any preference at all

**Comment by alikins** [comment]
> what do you mean by "wrapper mapping"? i do not have any preference at all

Not a particular thing, just a vague idea that the key id may need to be defined outside of the vault blob. One thought was something like a map with the (map) keys being 'key_id' and 'vault_blob', but I didn't really think that through ;->

Current version of my branch should be fine with multiline strings.

**Comment by alikins** [comment]
One advantage of keeping the decrypted value simple and deserializing to something string like is it makes it easier to wait until the last possible moment to decrypt (as opposed to attempting to decrypt while parsing). That could potentially be as late as remote node module execution[1]...

[1] assuming shared secrets or some pki support in ansible-vault...

**Comment by alikins** [comment]
@h0nIg 
Take a look at https://github.com/ansible/ansible/pull/16274

That is an implementation of vault encrypted yaml variables (originally based on your sample implementation)

**Comment by gebi** [comment]
Just as a note becaue i stumbled over this PR.
Speed would not be a problem if the wrong usage of PBKDF2 would be fixed in ansible vault (#20933).
The reason being, PBKDF2 should only be done ONCE per ansible startup to derive the possible low entropy user password into acceptable key material and not once per encrypted content.
The key derivation function done per content could be MUCH faster (eg. some simple HMAC with matching output length or PBKDF2 with only a single round instead of 10000).

the linked bug report is really about having a method to disable PBKDF2 with 10000 rounds for key material with already known high entropy and is no exact match to this discussion, but also discusses the principle problem in ansible vault.

**Comment by bcoca** [comment]
closing as implemented in #16274, released in 2.3 usable as `!vault` example i have above
